data/index.js

```diff
@@ -80,18 +137,65 @@ export function registerActions( reducerKey, newActions ) {
 }
 
 /**
- * Subscribe to changes to any data.
+ * Subscribe to changes to any data, optionally filtering by reducer or
+ * selector result.
  *
- * @param {Function}   listener Listener function.
+ * @param {?string}            reducerKey Reducer key.
+ * @param {?(string|string[])} selector   Selector name or array of selector
+ *                                        name with arguments to pass.
+ * @param {Function}           listener   Listener callback.
  *
- * @return {Function}           Unsubscribe function.
+ * @return {Function} Unsubscribe function.
  */
-export const subscribe = ( listener ) => {
-	listeners.push( listener );
+export const subscribe = ( ...args ) => {
```

Comment:
I like it a bit less, but not a blocker :) I can see the advantage of factorizing this kind of logic we could need in several places.

Reply:
The alternative being:

```js
let lastBlocks;
subscribe( () => {
	const blocks = select( 'core/editor' ).getBlocks();
	if ( blocks !== lastBlocks ) {
		// ... Blocks are known to have changed.
	}
	lastBlocks = blocks;
} );
```

Aside from being verbose and clumsy, there's another issue here: The selector is called on every single state change **to every single registered store**. For expensive queries, this would be needlessly expensive.

Reply:
https://github.com/WordPress/gutenberg/pull/5175/files#diff-180a3dfff6c915788f7e0b77d69b60c7R :)

Reply:
Right, but in your example the `subscribe` logic is run when `core/viewport` state changes, which is not so great.

Reply:
Minor not a blocker just a thought: In redux subscribes always receives a listener and there is no "filtering" possibility. In my opinion, I think we should also do the same, subscribe always receives just a listener, so we are consistent with redux. I understand filtering a specific change will be very a common case, so to address we can use: compose( selectorListner(selector), reducerListner( reducer ), subscribe ). This makes clear that some code is always executed for each change, and our functions selectorListner and reducerListner are just helpers to filter out change. Other helpers can be created in the future.

Reply:
@jorgefilipecosta This is interesting. At its most generic, it really only requires a "has value changed" higher-order function to encapsulate the behavior from the [original alternative](https://github.com/WordPress/gutenberg/pull/5206#discussion_r170108585). Something like:

```js
subscribe( ifChanged( () => (
	select( 'core/viewport' ).isViewportMatch( '< medium' )
) )( () => {
	const isViewportMatch = isViewportMatch( '< medium' );

	// Collapse sidebar when viewport shrinks.
	if ( isSmall ) {
		dispatch( 'core/edit-post' ).closeGeneralSidebar();
	}
} ) );
```

Then it becomes a question of: Do we expect this to be such a common pattern that it's still worth building conveniences into the API itself? The above snippet is generic, and not terribly verbose or clumsy, but still more verbose than the changes that had been proposed here. Generalism can be an ideal, but it shouldn't deter us from creating conveniences for common patterns.

It's a discussion worth having though:

- What other subscribe helpers could we imagine?
- What sorts of things would we expect people to use selector subscriptions for?

To the latter question, I'm finding that this could satisfy any need for a side effect pattern in the data module. In fact, taken to its extreme, we might be able to replace [all of our own internal side effects](https://github.com/WordPress/gutenberg/blob/master/editor/store/effects.js) with this approach, representing the side effects as a result of a change in state. Can all of the current and foreseen side effects be stated this way, or is access to the raw actions, even those not resulting in a change, necessary? It'd require some investigation, but I'm inclined to say that we could and should try to model side effects on changes in state. This might help remedy some of the disorganization that has resulted in the various `effects.js` files. At a higher-level, it might even be seen as the data equivalent of React's visual declarative UI, where "props" in this case are the selector result(s).

Reply:
As a testament to the clumsiness of https://github.com/WordPress/gutenberg/pull/5206#discussion_r170108585, this example code snippet inadvertently introduces an infinite loop 😏 (Dispatch causes subscribe to re-invoke before `lastBlocks` is assigned)

Reply:
Thank you for thoughts and analysing this, it is good that we will be able to discuss this specifc part outside of the scope of the rest of the changes 👍